### PR TITLE
Add Response::file_body

### PR DESCRIPTION
**Pull request by jbr**
This PR adds a method to tide::Response which incrementally improves sending a file with a content-size.

Before this change:
```rust
        app.at("/some_file").get(|_| async {
            let path = "./some_file.txt";
            let file = async_std::fs::File::open(path).await?;
            let size = file.metadata().await?.len();
            Response::new(StatusCode::Ok)
                .body(BufReader::new(file))
                .set_header(headers::CONTENT_LENGTH, size.to_string())
                .set_mime(mime_guess::from_path(path).first()?)
        });
```

After this change:
```rust
        app.at("/some_file").get(|_| async {
            let path = "./some_file.txt";
            let file = async_std::fs::File::open(path).await?;
            Response::new(StatusCode::Ok)
                .body_file(file).await?
                .set_mime(mime_guess::from_path(path).first()?)
        });
```

This PR also passes the length into the http client along with the body, allowing for optimizations or conditional logic based on the size. It would be really nice as a user of tide if body_file also guessed mime type, but this does not attempt to do that.

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
@jbr hey thanks for filing! I like the idea behind this patch, but it's got a few considerations to it:
1. It puts `async_std::fs::File` as part of Tide's public interface. This is the first time we'd do something like that, and I'm not sure if we'd want to. Perhaps it would need to live behind a feature flag?
2. It's currently the intention to move to [`http_types::Body`](https://docs.rs/http-types/1.2.0/http_types/struct.Body.html) as the constructor for body types, and then add `Response::set_body` to then set it. Through some generic conversion boundaries it would end up being quite ergonomic.

I think this may even be possible:

```rust
        app.at("/some_file").get(|_| async {
            let mut res = Response::new(StatusCode::Ok);
            res.set_body(File::open("./some_file.txt").await?);
            Ok(res)
        });
```

Perhaps the right way to go about this would be to add this behind an `"async-std"` feature flag in [`http_types::Body`](https://docs.rs/http-types/1.2.0/http_types/struct.Body.html) that's enabled by default. Then when we refactor the body constructors in Tide we can add simply forward the implementation as well.

Also worth noting is that for any stream in `http_types` we already do magic byte detection, so the mime type will be set already if it matches a known pattern. Perhaps we should expand that to match certain user-authored files as well (`{.js,.css,.html,.svg}` come to mind), but that could be another patch.

@jbr Would you be willing to make that first PR to `http-types`?

**Comment by prabirshrestha** [comment]
Coming from other webframeworks I find `body_file` confusing. Can we instead call it `send_file` that actually uses syscalls?

All major OS supports `sendfile` syscall which is highly optimized for sending files over sockets. I would personally prefer internally for this to be used.

- Linux [sendfile](http://man7.org/linux/man-pages/man2/sendfile.2.html)
- OSX [sendfile](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/sendfile.2.html)
- BSD [sendfile](https://www.freebsd.org/cgi/man.cgi?query=sendfile&sektion=2)
- Windows [TransmitFile](https://docs.microsoft.com/en-us/windows/win32/api/mswsock/nf-mswsock-transmitfile)

TLDR notes from the above docs:

linux docs:
>Because this copying is done within the kernel, sendfile() is more
       efficient than the combination of read(2) and write(2), which would
       require transferring data to and from user space

Windows docs:
>This function uses the operating system's cache manager to retrieve the file data, and provides high-performance file data transfer over sockets.

**Comment by jbr** [comment]
@yoshuawuyts I'll take a try at that PR :thumbsup:

As an alternative but not mutually exclusive approach suggestion, what do think about accepting an AsRef<Path> (or just an AsRef<str>) instead of the file itself, making it much more like https://hexdocs.pm/plug/Plug.Conn.html#send_file/5 — this would allow tide to make a mime guess based on file extension, allow the content-length determination used in this PR, and also _potentially_ in the distant future allow for the possibility of path-based content caching.

**Comment by yoshuawuyts** [comment]
On phone now, but Surf has a `body_file` method that takes a Path. I think that's the right argument (:

**Comment by jbr** [comment]
Addressed in https://github.com/http-rs/http-types/pull/106

**Comment by jbr** [comment]
Closing this because setting body will just pass through to http-types, if I understand the intent
